File: python3/delete_characters_to_make_fancy_string.py
"""
Space   : O(n)
Time    : O(n)
"""

class Solution:
    def makeFancyString(self, s: str) -> str:
        stack = []

        for l in s:
            stack.append(l)
            if len(stack) >= 3 and stack[-1] == stack[-2] ==  stack[-3] == l:
                stack.pop()

        return "".join(stack)

File: golang/delete_characters_to_make_fancy_string.go
package main

import "strings"

func makeFancyString(s string) string {
	stack := []string{}

	for _, l := range s {
		let := string(l)

		if len(stack) >= 2 {
			n := len(stack)
			if stack[n-1] == stack[n-2] && stack[n-1] == let {
				continue
			} else {
				stack = append(stack, let)
			}
		} else {
			stack = append(stack, let)
		}
	}

	return strings.Join(stack, "")
}
